core/src/test/scala/com/evolutiongaming/kafka/flow/RebalanceListenerSpec.scala

```diff
@@ -0,0 +1,125 @@
+package com.evolutiongaming.kafka.flow
+
+import cats.data.{NonEmptyMap, NonEmptySet, State}
+import cats.syntax.all._
+import com.evolutiongaming.catshelper.LogOf
+import com.evolutiongaming.kafka.flow.kafka.Consumer
+import com.evolutiongaming.kafka.journal.ConsRecords
+import com.evolutiongaming.skafka.consumer.{RebalanceListener => SRebalanceListener}
+import com.evolutiongaming.skafka.{Offset, OffsetAndMetadata, Partition, Topic, TopicPartition}
+import munit.FunSuite
+
+import scala.concurrent.duration.FiniteDuration
+
+class RebalanceListenerSpec extends FunSuite {
+
+  import RebalanceListenerSpec._
+
+  test("Listener should handle partitions assignment") {
+    val result = Fixture
+      .listener(topic1, topic2)
+      .onPartitionsAssigned(
+        NonEmptySet.of(
+          TopicPartition(topic1, partition1),
+          TopicPartition(topic1, partition2),
+          TopicPartition(topic2, partition3),
+          TopicPartition(topic2, partition4)
+        )
+      )
+      .runS(Context())
+      .value
+
+    val expected = List(
+      Action.Add(topic1, NonEmptySet.of(partition1 -> offset, partition2 -> offset)),
+      Action.Add(topic2, NonEmptySet.of(partition3 -> offset, partition4 -> offset))
+    )
+    assertEquals(result.actions.reverse, expected)
+  }
+
+  test("Listener should handle partitions revoke") {
+    val result = Fixture
+      .listener(topic1, topic2)
+      .onPartitionsRevoked(
+        NonEmptySet.of(
+          TopicPartition(topic1, partition1),
+          TopicPartition(topic2, partition3)
+        )
+      )
+      .runS(Context())
+      .value
+
+    val expected = List(
+      Action.Remove(topic1, NonEmptySet.of(partition1)),
+      Action.Remove(topic2, NonEmptySet.of(partition3))
+    )
+    assertEquals(result.actions.reverse, expected)
+  }
+
+  test("Listener should handle partitions lost") {
+    val result = Fixture
+      .listener(topic1, topic2)
+      .onPartitionsLost(
+        NonEmptySet.of(
+          TopicPartition(topic1, partition1),
+          TopicPartition(topic2, partition3)
+        )
+      )
+      .runS(Context())
+      .value
+
+    val expected = List(
+      Action.Remove(topic1, NonEmptySet.of(partition1)),
+      Action.Remove(topic2, NonEmptySet.of(partition3))
+    )
+    assertEquals(result.actions.reverse, expected)
+  }
+
+}
+
+object RebalanceListenerSpec {
+
+  val topic1 = "topic-1"
+  val topic2 = "topic-2"
+  val partition1 = Partition.unsafe(1)
+  val partition2 = Partition.unsafe(2)
+  val partition3 = Partition.unsafe(3)
+  val partition4 = Partition.unsafe(4)
+  val offset = Offset.unsafe(0)
+
+  type F[A] = State[Context, A]
+
+  implicit val logOf = LogOf.empty[F]
+
+  case class Context(actions: List[Action] = Nil) {
```

Comment:
I feel the test is good enough and elegant enough already to keep, but I would consider using "stub-style" instead of "mock-style" here.

I.e. store `Map[Topic, (Partition, Offset)]` instead of `List[Action]` and update that map instead.

Feel free to ignore this comment, if you feel "mock-style" test is better in this case as I have no strong opinion. The test already looks very cool.

See also https://martinfowler.com/articles/mocksArentStubs.html for inspiration.

Reply:
Thanks for pointing me on the article. I don't see any reasons why the test cannot be converted to "stub-style".

Done.